compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp

```diff
@@ -584,51 +832,119 @@ void PropagateLinalgTransposePass::runOnOperation() {
+                                                      reshapePropagationFn);
+    linalg::populateFoldReshapeOpsByCollapsingPatterns(bubblingPatterns,
+                                                       reshapePropagationFn);
+    linalg::FillOp::getCanonicalizationPatterns(bubblingPatterns, context);
+    bubblingPatterns.insert<BubbleTransposeThroughDpsInit>(context);
+    bubblingPatterns.add<PropagateTransposeThroughUnaryElementwiseDpsInit>(
+        context, /*benefit=*/2);
+    if (failed(applyPatternsAndFoldGreedily(funcOp,
+                                            std::move(bubblingPatterns)))) {
+      return signalPassFailure();
+    }
+  }
+
+  LLVM_DEBUG({
+    llvm::dbgs() << "\n--- After bubbling transpose ops up ---\n";
+    funcOp->print(llvm::dbgs(), OpPrintingFlags().useLocalScope());
+    llvm::dbgs() << "\n\n";
+  });
+
+  // Propagate transposes downwards, and fuse with any non-unary generic ops
+  // or linalg named ops. Also propagate reshapes downwards to open up more
+  // transpose fusion opportunities.
+  if (!testBubblingOnly) {
+    linalg::ControlFusionFn reshapePropagationFn = [](OpOperand *fusedOperand) {
+      Operation *producer = fusedOperand->get().getDefiningOp();
+      Operation *consumer = fusedOperand->getOwner();
+      if (!IREE::Flow::isNonNullAndOutsideDispatch({producer, consumer})) {
+        return false;
+      }
+      auto consumerLinalgOp = dyn_cast<linalg::LinalgOp>(consumer);
+      if (!consumerLinalgOp) {
+        return false;
+      }
+      // Only reshape generic ops and contraction ops.
+      if (!linalg::isaContractionOpInterface(consumerLinalgOp) &&
+          !isa<linalg::GenericOp>(consumerLinalgOp)) {
+        return false;
+      }
+      // Only propagate reshapes down through generics in this case, hoping
+      // to increase the chance we can fuse with a nearby transpose.
+      if (!isa<tensor::ExpandShapeOp>(producer) &&
+          !isa<tensor::CollapseShapeOp>(producer)) {
+        return false;
+      }
+      return isa_and_nonnull<linalg::TransposeOp>(
+          producer->getOperand(0).getDefiningOp());
+    };
+    RewritePatternSet sinkingPatterns(context);
+    linalg::populateFoldReshapeOpsByExpansionPatterns(sinkingPatterns,
+                                                      reshapePropagationFn);
+    linalg::populateFoldReshapeOpsByCollapsingPatterns(sinkingPatterns,
+                                                       reshapePropagationFn);
+    sinkingPatterns.insert<SinkTransposeThroughExtractSlice>(context);
+    sinkingPatterns.insert<SinkTransposeThroughExpandShape>(context);
+    sinkingPatterns.insert<FuseTransposeWithLinalgOpConsumer>(context);
+    populateNamedOpSinkingPatterns(context, sinkingPatterns);
+    sinkingPatterns.add<PropagateTransposeThroughUnaryElementwiseInput>(
+        context, /*benefit=*/2);
+
     if (enableAggressivePropagation) {
       sinkingPatterns.insert<GeneralizeInputTransposedNamedOp>(context);
```

Comment:
You can use a similar trick for aggressive bubbling by having a pattern that looks at the consumer of the named op and generalizes it if there is a transpose.